抽出したハンドラー関数でSearchFormのJSXを簡素化

toshinori-m opened the pull request:
### 概要
- セレクトボックスにて`SearchForm` を利用しコンポーネント内のロジックを、関数としてjsx外に抽出しました。

### 変更内容
- セレクトボックスを`SearchForm` を利用し共通化
- JSX内に直接書かれていた `onChange` ロジックを分離
- フォームの可読性・保守性が向上

### 動作確認
既存の動作に影響がないことを確認済み。

close https://github.com/toshinori-m/stay_connect/issues/201

toshinori-m committed:
refactor: 抽出したハンドラー関数でSearchFormのJSXを簡素化

hirosi1900day reviewed (approved):
ここもuseActionStateを使う形にできるかもですが変更自体は問題なさそう

toshinori-m commented:
@hirosi1900day 
はい、useActionStateを使う形にしたいと思っていますが、下のprのマージがまだなのでuseActionStateの使用方法（form -> formData -> バリデーション -> apiリクエスト）にしようか迷っている状況です！
・[イベント設定ページに開始・終了日付、募集チーム数、目的・その他入力欄を追加 ](https://github.com/toshinori-m/stay_connect/pull/199)

toshinori-m committed:
refactor: useActionStateを使う形に改善

toshinori-m commented:
useActionStateを使う形に改善してみました。

よろしくお願いします

hirosi1900day commented on frontend-react/src/components/HomePage/SearchForm.tsx:
```diff
@@ -1,73 +1,101 @@
-              ))}
-            </select>
-          ) : null
+      <form onSubmit={handleSubmit} className="my-5 text-center">
```
useActionStateを使う形になってます？？

toshinori-m replied:
申し訳ありません(^_^;)

toshinori-m committed:
refactor: useActionStateを使う形に改善（再度）

hirosi1900day commented on frontend-react/src/components/HomePage/SearchForm.tsx:
```diff
@@ -1,73 +1,115 @@
+    await onSearch()
+
+    // 検索実行後、状態を復元
+    setTimeout(() => {
```
このsetTimeoutはなんのために必要なのでしょうか？
(setTimeoutは基本的にこういう場面では使わない方が良いかなと思います)

toshinori-m replied:
@hirosi1900day 
この `setTimeout` を入れている理由は、検索実行直後にフォーム状態が一度リセットされることを防ぐために使用しました。

例えば下記のような流れです。
1. 「競技」で「サッカー」を選択
2. 検索ボタンを押す（→この時、選択状態が一時的に「競技選択」に戻る）
3. 別の項目（例：都道府県）を変更しようとすると、「競技」が再び「サッカー」に戻る

他に良い方法があればぜひ教えて下さい

hirosi1900day replied:
そういった用途であれば最悪useRefを使う形になるかなと思いますが、以下のようにformDataから取得しステートにセットし直すではダメなんでしたっけ？？(少なくともsetTimeoutで対応するのは適切ではないと思ってます）

https://github.com/toshinori-m/stay_connect/blob/6b5ba90a256ef1471e62fbd075b5fa24ce8e28b9/frontend-react/src/pages/eventPages/EventSettingPage.tsx#L96-L104

hirosi1900day commented on frontend-react/src/pages/HomePage.tsx:
```diff
@@ -121,19 +141,14 @@ export default function HomePage() {
     <div className="mt-32 md:mt-20 mx-auto p-4 md:flex md:items-start">
       <SearchForm
```
SearchFormにpropsでロジックやstateを渡す設計となってますが、コードを読む際にHomePageコンポーネントとsearchFormコンポーネントを行ったり来たりする必要があり、かなり読みづらいので
frontend-react/src/components/HomePage/SearchForm.tsxにロジック等全て持っていけば高凝集になり読みやすくなる気がしますがいかがでしょうか？

(対応はこのPR出なくても良いです)

toshinori-m replied:
このHOME画面は次に下のissueに対応する必要があるので、本PRで対応させて下さい

[HOME画面のイベント選択にてイベント画面表示](https://github.com/toshinori-m/stay_connect/issues/214)

toshinori-m committed:
Refactor: 以下のリファクタリングを実施

- 検索エラー表示処理を HomePage から SearchForm に移管
- コンポーネントの責務を明確化し、可読性を向上
- setTimeout を用いた状態復元処理を削除
- formData から直接 state を再構築する設計に変更

hirosi1900day commented on frontend-react/src/pages/HomePage.tsx:
```diff
@@ -15,96 +14,36 @@ interface Recruitment {
-  const [errors, setErrors] = useState<string[]>([])
   const apiClient = useApiClient()
+  const [recruitments, setRecruitments] = useState<Recruitment[]>([])
+  const [searchErrors, setSearchErrors] = useState<string[]>([])
```
このあたりもsearchFormに持っていけそう

hirosi1900day commented on frontend-react/src/pages/HomePage.tsx:
```diff
@@ -15,96 +14,36 @@ interface Recruitment {
-    fetchSportsTypes()
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, [sportsTypeSelected])
+  const handleSearch = async (searchParams: Record<string, string>) => {
```
このあたりもsearchFormに持っていけそう

hirosi1900day commented on frontend-react/src/pages/HomePage.tsx:
```diff
@@ -119,22 +58,7 @@ export default function HomePage() {
-      />
+      <SearchForm onSearch={handleSearch} externalErrors={searchErrors} />
 
       <div className="md:w-5/6 md:ml-2">
```
このブロックもsearchFormに持っていけそうな気がする and searchFormの責務といっても問題ない気がするですが
HomePageに持たせておかないといけない理由とかあったりしますか？
この辺りがsearchFormに行けばrecruitmentsのstate管理も一つのコンポーネントに閉じることができるのでかなりみやすくなりかなと思ったのですがいかかがでしょうか？
(pageコンポーネントは、各featureコンポーネントをまとめて呼び出すだけにとどめ、ロジックは極力持たない構成が理想です。
例えば、HomePage.tsでは、SearchFormなどのコンポーネントを組み合わせて表示するだけにし、ロジックはfeatureコンポーネント側に集約するようにします。

こうすることで、機能ごとの責務が明確になり、各コンポーネント単位でテストを実施しやすくなります。結果として、保守性の高い構成になります。
もちろん、現実的には完全に分離しきるのは難しいこともありますが、目指すべき方向性としてはこの形が望ましいです。

.)

toshinori-m replied:
HomePage.tsx はHome画面を表示するjsxを中心にし、SearchForm.tsx は検索フォーム周りのロジックや状態管理を中心となるように修正しました。
修正前は検索フォームと検索結果を別々にした方が責務が分かりやすいかと思っていたのですが、今回の構成の方がよい方向でしょうか？
また、他のページでもこの構成をベースにしていく感じになりますか？

hirosi1900day replied:
どちらかというとHomePage.tsx に書くのではなく、SearchForm.tsx なの中に検索フォームとその結果表示を入れちゃってもいいかなと思ってましたが、修正前の形の方がよければその形でも良いかもしれないですね(好みの問題もあるので)

自分がコンポーネントに切り出す単位としてはテストが完結する単位で切ることが多いのでformと表示がセットで1機能なのであればその単位で切り出してしまっても良いかなと思ったというのが以下のご質問の意図です。
(いろんなファイルに行ったり来たりをなん度もしないと読めないコードでなければOKだと思います)

> このブロックもsearchFormに持っていけそうな気がする and searchFormの責務といっても問題ない気がするですが
HomePageに持たせておかないといけない理由とかあったりしますか？
この辺りがsearchFormに行けばrecruitmentsのstate管理も一つのコンポーネントに閉じることができるのでかなりみやすくなりかなと思ったのですがいかかがでしょうか？
(pageコンポーネントは、各featureコンポーネントをまとめて呼び出すだけにとどめ、ロジックは極力持たない構成が理想です。
例えば、HomePage.tsでは、SearchFormなどのコンポーネントを組み合わせて表示するだけにし、ロジックはfeatureコンポーネント側に集約するようにします。

HomePage.tsxは以下だけの責務
```
return (
  </ 今後追加するかもしれないコンポーネント>
  </SearchForm>
  </ 今後追加するかもしれないコンポーネント>
)
```

toshinori-m replied:
合っているのかわかりませんが修正してみました。

よろしくお願いします。

toshinori-m committed:
Refactor: HomePage 用の検索フォームロジックを hooks/useSearchForm に抽出

toshinori-m committed:
refactor: 検索フォームと結果表示をSearchForm.tsxに統合し、機能単位で完結する構成に変更
